imap_processing/glows/l1b/glows_l1b.py

```diff
@@ -0,0 +1,142 @@
+"""Methods for processing GLOWS L1B data."""
+
+import dataclasses
+
+import xarray as xr
+
+from imap_processing.glows.l1b.glows_l1b_data import DirectEventL1B, HistogramL1B
+
+
+def glows_l1b_histograms(input_dataset: xr.Dataset) -> xr.Dataset:
+    """Process the GLOWS L1B data and format the histogram attribute outputs."""
+    # TODO: add CDF attribute steps
+
+    # Will call process_histograms and construct the dataset with the return
+    raise NotImplementedError
+
+
+def glows_l1b_de(input_dataset: xr.Dataset) -> xr.Dataset:
+    """Process the GLOWS L1B data and format the direct event attribute outputs."""
+    # TODO: generate dataset with CDF attributes
+    # Will call process_de and construct the dataset with the return
+    raise NotImplementedError
+
+
+def process_de(l1a: xr.Dataset) -> tuple[xr.DataArray]:
+    """
+    Process the direct event data from the L1A dataset and return the L1B dataset.
+
+    This method takes in a dataset and applies a function along each timestamp.
+    The function here is the creation of a DirectEventL1B object from the data. Each
+    variable is passed in as an argument to the lambda function, and then a
+    DirectEventL1B object is created from those arguments. This way, each timestamp
+    gets its own DirectEventL1B object.
+
+    Parameters
+    ----------
+    l1a : xr.Dataset
+        The L1A dataset to process
+
+    Returns
+    -------
+    l1b_arrays: tuple[xr.DataArray]
+        The DataArrays for each variable in the L1B dataset
+    """
+    dataarrays = [l1a[i] for i in l1a.keys()]
+
+    # Set the dimensions for the input and output dataarrays
+    # The dimension include all the non-epoch dimensions. Epoch is the dimension that
+    # the data is processed along - so the data from *dataarrays for each timestamp
+    # is passed into the function (here a lambda.)
+
+    # We need to specify the other dimensions for input and output so the arrays are
+    # properly aligned. The input dimensions are in `dims` and the output dimensions are
+    # in `new_dims`.
+
+    # An empty array passes the epoch dimension through
+    dims = [[] for i in l1a.keys()]
+    new_dims = [[] for i in range(14)]
+
+    # Set the two direct event dimensions. This is the only multi-dimensional L1A
+    # (input) variable.
+    dims[0] = ["per_second", "direct_event"]
+
+    # Flags is a constant length. It has a dimension of "flags"
+    new_dims[-3] = ["flags"]
+
+    # glows_times and pulse_lengths should be dimension of "per_second", the same as
+    # the input.
+    new_dims[-2] = ["per_second"]
+    new_dims[-1] = ["per_second"]
+
+    l1b_fields = xr.apply_ufunc(
+        lambda *args: tuple(dataclasses.asdict(DirectEventL1B(*args)).values()),
```

Comment:
I'm not sold on this style personally, but again I'm happy to be the odd one out. I'd find this much easier to follow to use typical numpy-style broadcasting on things and I think it would be much faster because you'd iterate in C rather than iterate over each time via dataclasses. (I also think this might be thrashing disk I/O opening the params file for each iteration)

Just to be explicit, something like this is what I would personally find easier follow with processing functions being applied rather than class-oriented methods.
```python
def decode_ds(ds, param_key, encoded_key):
        params = ds[param_key]
        # compute parameters a and b:
        param_a = (2 ** params["n_bits"] - 1) / (params["max"] - params["min"])
        param_b = -params["min"] * param_a

        return (ds["encoded_key"] - param_b) / param_a).astype(np.single)

# Then call it on the dataarrays of interest
output_data_array = decode_ds(l1a_ds, "hv_voltage", "hv_voltage_encoded")
...

# Or apply this "decode" function to the dataset and let the numpy vectorization take place there without needing to go to dataclasses at each epoch timestep.
output_data_array = xr.apply_ufunc(decode_ds, l1a_ds, ...)
```

Reply:
You're right about the I/O, I need to move that outside of the class anyway. Personally, I prefer the way it's organized in the dataclass rather than having loads of functions, but I can see the advantages of both. That being said, I don't actually think either method is much faster or slower (excluding the IO thing) because of the way `apply_ufunc` [is written](https://docs.xarray.dev/en/stable/generated/xarray.apply_ufunc.html). 

If I wanted it to be faster, I would need to write my own pre-vectorized functions so I could get that advantage from C, but as it is now, I am pretty sure it's doing similar amounts of work either way. In any case, I think this is still faster than what most of the other instruments are using or my previous designs, because it is guaranteed to only loop through the data one time. It doesn't need to loop through again to set up the datasets or to rearrange the dimensions. Plus, most of the real work inside the methods does take advantage of those numpy vectorizations that give speed. 

I prefer having it organized in a class so it's super explicit what the output is expected to be, especially in for example the direct events class where attributes are dropped or combined. I think it would be very confusing in a function call and difficult to tell at a glance what the actual output is.

Reply:
I did a quick profiler test with a simple example of both, and here are the results for the method: 
` 6438 function calls (6294 primitive calls) in 0.004 seconds`

vs the dataclass:
`6648 function calls (6495 primitive calls) in 0.004 seconds`

So, the dataclass might actually be a hair faster, but it's certainly comparable as long as the methods are well-written.

Reply:
Thanks for checking! Can you paste the quick example script?

> I prefer having it organized in a class so it's super explicit what the output is expected to be, especially in for example the direct events class where attributes are dropped or combined. I think it would be very confusing in a function call and difficult to tell at a glance what the actual output is.

👍 Just a difference of opinion, so that sounds good to me. I find functions easier to interpret at first glance 🤷

Reply:
```
import dataclasses
from dataclasses import dataclass

import xarray as xr
import numpy as np

import cProfile

def foo(test_dataset):
    @dataclass
    class TestClass:
        direct_events: np.ndarray
        support_data_1: float
        support_data_2: float

        def __post_init__(self):
            self.support_data_1 = self.support_data_1 + 1
            self.support_data_2 = self.support_data_2 * 3

    def mapping(direct_events, support_data_1, support_data_2):
        result1 = direct_events
        result2 = support_data_1 + 1
        result3 = support_data_2 * 3
        return result1, result2, result3

    # Input and output core dims should include all dims EXCEPT Epoch
    # de_out, data_1_out, data_2_out = xr.apply_ufunc(
    #     mapping,
    #     test_dataset["direct_events"],
    #     test_dataset["support_data_1"],
    #     test_dataset["support_data_2"],
    #     input_core_dims=[["per_seconds"], [], []],
    #     output_core_dims=[["per_seconds"], [], []],
    #     vectorize=True,
    # )

    de_out2, data_1_out2, data_2_out2 = xr.apply_ufunc(
        lambda *args: tuple(dataclasses.asdict(TestClass(*args)).values()),
        test_dataset["direct_events"],
        test_dataset["support_data_1"],
        test_dataset["support_data_2"],
        input_core_dims=[["per_seconds"], [], []],
        output_core_dims=[["per_seconds"], [], []],
        vectorize=True,
    )

direct_events = np.array([[[1, 2, 3, 4], [5, 6, 7, 8]],
                              [[9, 10, 11, 12], [13, 14, 15, 16]],
                              [[17, 18, 19, 20], [21, 22, 23, 24]]])

direct_events_2 = np.array([[1, 2, 3, 4],
                            [9, 10, 11, 12],
                            [17, 18, 19, 20]])

support_data_1 = np.array([-2, -3, -4])
support_data_2 = np.array([1, 2, 3])
print(direct_events.shape)
time_data = np.array([100, 101, 102])

epoch = xr.DataArray(
    time_data,
    name="epoch",
    dims=["epoch"]
)

direct_event = xr.DataArray(
    np.arange(4),
    name="direct_event",
    dims=["direct_event"])

per_seconds = xr.DataArray(
    np.arange(4),
    name="per_seconds",
    dims=["per_seconds"])

de = xr.DataArray(
    direct_events_2,
    name="direct_events",
    dims=["epoch", "per_seconds"],
    coords={"epoch": epoch, "per_seconds": per_seconds})

test_dataset = xr.Dataset(
    coords={"epoch": time_data}
)

test_dataset["direct_events"] = de
test_dataset["support_data_1"] = xr.DataArray(
    support_data_1,
    name="support_data_1",
    dims=["epoch"],
    coords={"epoch": epoch}
)

test_dataset["support_data_2"] = xr.DataArray(
    support_data_2,
    name="support_data_2",
    dims=["epoch"],
    coords={"epoch": epoch}
)


cProfile.run('foo(test_dataset)')
```

Reply:
That's awesome @maxinelasp, thanks for providing it. I'm always a little leary of small array profiles. Here is a quick modification with the takeaway being that I'm still not sold on this approach, but I'm not going to block either.

```
# cProfile.run('foo_dataclass(test_dataset)')  # 7601328 function calls (7301312 primitive calls) in 1.383 seconds
# cProfile.run('foo_xarray_ufunc(test_dataset)')  # 901328 function calls (901312 primitive calls) in 0.172 seconds
# cProfile.run('foo_direct_call(test_dataset)')  # 278 function calls (276 primitive calls) in 0.000 seconds
```

```python
import dataclasses
from dataclasses import dataclass

import xarray as xr
import numpy as np

import cProfile

@dataclass
class TestClass:
    direct_events: np.ndarray
    support_data_1: float
    support_data_2: float

    def __post_init__(self):
        self.support_data_1 = self.support_data_1 + 1
        self.support_data_2 = self.support_data_2 * 3

def mapping(direct_events, support_data_1, support_data_2):
    result1 = direct_events
    result2 = support_data_1 + 1
    result3 = support_data_2 * 3
    return result1, result2, result3

def foo_dataclass(test_dataset):
    return xr.apply_ufunc(
        lambda *args: tuple(dataclasses.asdict(TestClass(*args)).values()),
        test_dataset["direct_events"],
        test_dataset["support_data_1"],
        test_dataset["support_data_2"],
        input_core_dims=[["per_seconds"], [], []],
        output_core_dims=[["per_seconds"], [], []],
        vectorize=True,
    )

def foo_xarray_ufunc(test_dataset):
    # Input and output core dims should include all dims EXCEPT Epoch
    return xr.apply_ufunc(
        mapping,
        test_dataset["direct_events"],
        test_dataset["support_data_1"],
        test_dataset["support_data_2"],
        input_core_dims=[["per_seconds"], [], []],
        output_core_dims=[["per_seconds"], [], []],
        vectorize=True,
    )


def foo_direct_call(test_dataset):
    return mapping(
        test_dataset["direct_events"],
        test_dataset["support_data_1"],
        test_dataset["support_data_2"]
    )

ntimes = 100000
nevents = 400
per_seconds = 30

direct_events = np.arange(ntimes*nevents*per_seconds).reshape(ntimes, nevents, per_seconds)
direct_events_2 = np.arange(ntimes*per_seconds).reshape(ntimes, per_seconds)

support_data_1 = np.arange(ntimes)
support_data_2 = np.arange(ntimes)
print(direct_events.shape)
time_data = np.arange(ntimes)

epoch = xr.DataArray(
    time_data,
    name="epoch",
    dims=["epoch"]
)

direct_event = xr.DataArray(
    np.arange(nevents),
    name="direct_event",
    dims=["direct_event"])

per_seconds = xr.DataArray(
    np.arange(per_seconds),
    name="per_seconds",
    dims=["per_seconds"])

de = xr.DataArray(
    direct_events_2,
    name="direct_events",
    dims=["epoch", "per_seconds"],
    coords={"epoch": epoch, "per_seconds": per_seconds})

test_dataset = xr.Dataset(
    coords={"epoch": time_data}
)

test_dataset["direct_events"] = de
test_dataset["support_data_1"] = xr.DataArray(
    support_data_1,
    name="support_data_1",
    dims=["epoch"],
    coords={"epoch": epoch}
)

test_dataset["support_data_2"] = xr.DataArray(
    support_data_2,
    name="support_data_2",
    dims=["epoch"],
    coords={"epoch": epoch}
)


a1, b1, c1 = foo_dataclass(test_dataset)
a2, b2, c2 = foo_xarray_ufunc(test_dataset)
a3, b3, c3 = foo_direct_call(test_dataset)

# Verify all functions provide same results
np.testing.assert_array_equal(a1, a2)
np.testing.assert_array_equal(a1, a3)
np.testing.assert_array_equal(b1, b2)
np.testing.assert_array_equal(b2, b3)
np.testing.assert_array_equal(c1, c2)
np.testing.assert_array_equal(c2, c3)

# cProfile.run('foo_dataclass(test_dataset)')  # 7601328 function calls (7301312 primitive calls) in 1.383 seconds
# cProfile.run('foo_xarray_ufunc(test_dataset)')  # 901328 function calls (901312 primitive calls) in 0.172 seconds
# cProfile.run('foo_direct_call(test_dataset)')  # 278 function calls (276 primitive calls) in 0.000 seconds
```

Reply:
I investigated your code a bit more, and turns out a lot of the slowdown from the dataclass is from the `dataclasses.asdict` method, which makes a deep copy. If I just manually write a function to return the tuple without a deepcopy: 

```
def output(self):
  return tuple(getattr(self, field.name) for field in dataclasses.fields(self))
```

Then I get something a lot closer to the direct function call: 

`2601395 function calls (2601379 primitive calls) in 0.448 seconds`

This is also a fixed slow down, so if the function calls are slower, the difference between the two methods stays at ~0.3s (which I checked by just adding a load of factorial calls in both methods)

My instincts are that they should be a lot closer together than that, so I think the slowdown is coming from inside the dataclass implementation. But, I still think it's worth the trade off for a very very minor performance hit to have a class which perfectly describes the output file and has docs right there for every attribute.

Also, thank you for the nerd snipe, I was supposed to be doing things today 😂 instead I was figuring out where that difference came from.

Reply:
:) Glad you're enjoying doing the deep dive!

> But, I still think it's worth the trade off for a very very minor performance hit to have a class which perfectly describes the output file and has docs right there for every attribute.

👍 That is fine, I just have the opposite opinion and find it more confusing to track that documentation than through the function calls, but no worries this is your instrument!

Note that the direct function call is actually the one I'd prefer, avoiding `xr.apply_ufunc` altogether. That is still more than 100x faster (maybe more because I'm just getting 0.000s).